### PR TITLE
include: linker: Fix linker sections

### LINKED ISSUE #12490: Produced ELF does not follow the linux ELF spec
**Describe the bug**
Tried to ask this on slack but got no answer. I've a question about the output of the linker-script and elf for ARM. Why does it not follow the Linux ELF spec/convention for the section headers when it comes to text, bss etc. but does so for `.comment`?

[SPEC](http://refspecs.linuxbase.org/elf/elf.pdf ) As the output of the section headers in the ELF produced does not follow the spec and also outputs a _TEXT_SECTION_NAME_2 section header which I assume is not supposed to be there.

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=\<an ARM board>
3. make
4. `readelf -S main \<builddir>/zephyr/zephyr.elf

**Expected behavior**
To follow the ELF spec and by then having a `.` before the section names described in `include/linker/sections.h`

**Impact**
Mostly annoyance but also makes including other parts into the linker-script with `custom-<type>.ld` a bit harder.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK
 - Master

**Additional context**
This is an open question I'll submit a PR with a fix for the sections named specified by the SPEC but looking at the project this could be a design decision so I'm mostly looking for answeres.

### PR DESCRIPTION
**Pull request by sigvartmh**
Changes the linker sections to follow the
[ELF Spec](http://refspecs.linuxbase.org/elf/elf.pdf)
instead of using sections without `.` infront for special
sections.

Fixes #12490

Signed-off-by: Sigvart M. Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

**Review by SebastianBoe — APPROVED** [review]

File: include/arch/arm/cortex_m/scripts/linker.ld
Comment (SebastianBoe) [inline include/arch/arm/cortex_m/scripts/linker.ld]: _TEXT_SECTION_NAME_ isn't a token recognized by CPP, so this doesn't work, does it?

Should be _TEXT_SECTION_NAME

Comment (sigvartmh) [reply include/arch/arm/cortex_m/scripts/linker.ld]: It works for GCC 6.2 for C but I'm not sure about CPP anyways changing it to _<NAME> instead does no harm. I'll change it

Comment (SebastianBoe) [reply include/arch/arm/cortex_m/scripts/linker.ld]: Sorry, I meant the C preprocessor, not C plus plus.

Strange, I wouldn't have expected this to work.

File: include/linker/sections.h
Comment (SebastianBoe) [inline include/linker/sections.h]: After having a second look. I suspect we want to use a different output section name for the first text section. E.g. .init.

And then use .text for the "main" text section.

This would mean that additional declarations of output sections for .text would go into the right place. And this would match the users expectation of what is in .text .

Comment (sigvartmh) [reply include/linker/sections.h]: Do agree I'll draft up a proposal

**Comment by sigvartmh** [comment]
```
There are 28 section headers, starting at offset 0xcbe40:

Section Headers:
  [Nr] Name              Type            Addr     Off    Size   ES Flg Lk Inf Al
  [ 0]                   NULL            00000000 000000 000000 00      0   0  0
  [ 1] .init             PROGBITS        00000000 0000b8 000100 00 WAX  0   0  4
  [ 2] .text             PROGBITS        00000100 0001b8 007e6c 00 WAX  0   0  8
  [ 3] .ARM.exidx        ARM_EXIDX       00007f6c 008024 000008 00  AL  2   0  4
  [ 4] sw_isr_table      PROGBITS        00007f74 00802c 000180 00  WA  0   0  4
  [ 5] devconfig         PROGBITS        000080f4 0081ac 000084 00  WA  0   0  4
  [ 6] .rodata           PROGBITS        00008178 008230 000614 00   A  0   0  4
  [ 7] .bss              NOBITS          20000000 008b28 000ea8 00  WA  0   0  8
  [ 8] .noinit           NOBITS          20000ea8 008b28 0018f8 00  WA  0   0  8
  [ 9] .data             PROGBITS        200027a0 008848 0001e0 00  WA  0   0  8
  [10] initlevel         PROGBITS        20002980 008a28 000084 00  WA  0   0  4
  [11] _k_sem_area       PROGBITS        20002a04 008aac 000018 00  WA  0   0  4
  [12] _k_mutex_area     PROGBITS        20002a1c 008ac4 000014 00  WA  0   0  4
  [13] _k_queue_area     PROGBITS        20002a30 008ad8 000010 00  WA  0   0  4
  [14] _net_buf_pool_are PROGBITS        20002a40 008ae8 000040 00  WA  0   0  4
  [15] .comment          PROGBITS        00000000 008b28 000075 01  MS  0   0  1
  [16] .debug_aranges    PROGBITS        00000000 008ba0 001880 00      0   0  8
  [17] .debug_info       PROGBITS        00000000 00a420 057035 00      0   0  1
  [18] .debug_abbrev     PROGBITS        00000000 061455 00e90d 00      0   0  1
  [19] .debug_line       PROGBITS        00000000 06fd62 01e3bc 00      0   0  1
  [20] .debug_frame      PROGBITS        00000000 08e120 003aa4 00      0   0  4
  [21] .debug_str        PROGBITS        00000000 091bc4 00bc86 01  MS  0   0  1
  [22] .debug_loc        PROGBITS        00000000 09d84a 01d126 00      0   0  1
  [23] .debug_ranges     PROGBITS        00000000 0ba970 005268 00      0   0  8
  [24] .ARM.attributes   ARM_ATTRIBUTES  00000000 0bfbd8 000031 00      0   0  1
  [25] .symtab           SYMTAB          00000000 0bfc0c 0074a0 10     26 1164  4
  [26] .strtab           STRTAB          00000000 0c70ac 004c6a 00      0   0  1
  [27] .shstrtab         STRTAB          00000000 0cbd16 000128 00      0   0  1
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  y (purecode), p (processor specific)
```
The ELF section headers look like this with this patch. The `sanitycheck` script still needs to be updated with the correct names for the sections and the additional `.init` section needs to be added.

**Comment by carlescufi** [comment]
@sigvartmh is this ready to go in or do we need to update the sanitycheck script?

**Comment by sigvartmh** [comment]
@carlescufi not ready, the sanitycheck script needs to be updated

**Comment by nashif** [comment]
reasoning for the DNM? this is not your PR and if you do not want this to be merged, you should review and give it a -1 instead.

**Review by carlescufi — APPROVED** [review]

**Comment by carlescufi** [comment]
@andrewboie can you please review this one?

**Comment by sigvartmh** [comment]
#12490 pushing some information up to this PR https://github.com/zephyrproject-rtos/zephyr/blob/128d2e2bd176b8c5e384ea46ea452e7864cce84f/scripts/sanitycheck#L674 needs to be updated before this is merged I think. I'm not familiar with what the intention of this script is but from the class name it seems it tries to calculate some sizes. By updating the linker script these calculations would be off. A naive way of doing it is just doing a "search and replace" with the renaming of the sections but since I'm not familiar with the intention of the script this could break stuff.

**Comment by ioannisg** [comment]
Shall we try to get this PR is @carlescufi @sigvartmh ? Is this still applicable?

**Comment by aescolar** [comment]
The `DNM` label was due to these comments [1](https://github.com/zephyrproject-rtos/zephyr/pull/12510#issuecomment-458627543) [2](https://github.com/zephyrproject-rtos/zephyr/pull/12510#issuecomment-461024524 ). I removed the label now. @sigvartmh  please add/remove it as you see fit.

File: include/linker/sections.h
Comment (andrewboie) [inline include/linker/sections.h]: you could call this .app_data

**Review by tyrantlizard77 — APPROVED** [review]

File: include/linker/sections.h
Comment (ioannisg) [reply include/linker/sections.h]: Yes, I would like that as well.

**Review by ioannisg — CHANGES_REQUESTED** [review]
Fix commit messages so they correspond to what is actually changed in each patch

**Comment by sigvartmh** [comment]
Sorry guys haven't seen this before now. I'll fix it up, not sure if it is still relevant or stale tho?

**Pull request closed without merging**
